### PR TITLE
feat: flexible commands in issue comments

### LINKED ISSUE #46: issue comment commands should offer more options
Currently we have `/run bisect` which works but only works in one specific way, e.g. you need to know exactly how to format an issue template to provide the good & bad versions for bisection. If you miss any part, nothing will happen.

Also, we might want to make it more explicit that bugbot is being invoked, in the same way we have `/trop ${action} ${args...}` maybe we should support options in the command.

- `/bugbot bisect $gist $good $bad`
- `/bugbot test $gist [$version]`

### LINKED ISSUE #72: in-comment commands do not get activated when surrounded by whitespace
This command will trigger a bisect:

```
/test bisect
```

But this will not:

```
/test bisect

```

### PR DESCRIPTION
**Pull request by ckerr**
Fixes #46.

Fixes #72.

Update the probot client's issue-parser so that it can return different command types. Previously it only did bisection requests; now we need test requests too.

The new syntax:

 `/bugbot bisect [gistId] [goodVersion [badVersion]]`

- If no `gistId` is given, use the issue body's `Testcase Gist URL`.
- If no `goodVersion` is given, use the issue body's `Last Known Working Electron Version` or an old version.
- If no `badVersion` is given, use the issue body's `Electron Version` or the latest release.

`/bugbot test [gist] [platforms...] [versions...]`

- If no `gist` is given, use the issue body's `Testcase Gist URL`.
- If no `platforms` are given, test on Linux, Mac, and Windows.
- If no `versions` are given, test the first version and latest version of each prerelease branch, of each supported branch, and of the two branches before that.

This is all explained now at the top of the monorepo README.md so that we can link to it / so that it will be easier for maintainers to find.

### CODE REVIEW

File: modules/bot/test/fixtures/issue-gist-trailing-slash.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ x64
 
 ### Last Known Working Electron Version
 
-13.0.0
+12.0.0
```
Comment (ckerr) [inline modules/bot/test/fixtures/issue-gist-trailing-slash.md:25]: Side comment: fixed wart in the issue-parser test + fixtures that specified the good release to be newer than the bad release

File: modules/bot/src/github-client.ts [hunk 2]
```diff
@@ -186,6 +154,20 @@ export class GithubClient {
...
+   * try {
+   *   currentJob = await this.broker.getJob(id);
+   * } catch (e) {
+   *    // no-op
+   * }
+   * if (action === actions.STOP && currentJob && !currentJob.time_finished) {
+   *   this.broker.stopJob(id);
+   * } else if (action === actions.BISECT && !currentJob) {
+   */
+
```
Comment (clavin) [inline modules/bot/src/github-client.ts:170]: Is this meant to be left here? In the middle of the class? If so, what is this a draft implementation *of*?

File: modules/bot/src/issue-parser.ts [hunk 3]
```diff
@@ -1,36 +1,55 @@
...
+  }
+
+  return sections;
 }
 
 // copied from Fiddle code
 export function getGistId(input: string): string | null {
   let id: string | undefined = input;
-  if (input.startsWith('https://gist.github.com/')) {
+  if (input?.startsWith('https://gist.github.com/')) {
```
Comment (clavin) [inline modules/bot/src/issue-parser.ts:52]: The type `string` can't be `null` or `undefined` though?

File: modules/bot/src/issue-parser.ts [hunk 4]
```diff
@@ -42,34 +61,145 @@ export function getGistId(input: string): string | null {
...
   }
 
-  const gistId = getGistId(gistUrl);
+  // if any pieces are missing, fill them in from the issue body
+  const sections = splitMarkdownByHeader(issueBody);
+  d('sections', inspect(sections));
+  badVersion ||= SemVer.coerce(sections.get(BAD_VERSION))?.version;
+  badVersion ||= await versions.getLatestVersion();
+  goodVersion ||= SemVer.coerce(sections.get(GOOD_VERSION))?.version;
+  goodVersion ||= await versions.getDefaultBisectStart();
```
Comment (clavin) [inline modules/bot/src/issue-parser.ts:106]: long nit:

I think `version ??= ... ?? ...;` may be more clear than `version ||= ...; version ||= ...;`

Alternatively, name the two previous variables `argBadVersion` and `argGoodVersion`, make their types `string | undefined`, and then make these two statements sort of like this:

```ts
const badVersion = argBadVersion ?? ... ?? ...;
// (and then again for goodVersion)
```

File: modules/bot/src/issue-parser.ts [hunk 5]
```diff
@@ -42,34 +61,145 @@ export function getGistId(input: string): string | null {
...
+    [goodVersion, badVersion] = [badVersion, goodVersion];
+  }
+
+  d({ badVersion, gistId, goodVersion });
+  return badVersion && gistId && goodVersion
+    ? { type: 'bisect', badVersion, gistId, goodVersion }
+    : undefined;
+}
+
+const allPlatforms = ['darwin', 'linux', 'win32'];
```
Comment (clavin) [inline modules/bot/src/issue-parser.ts:125]:
```suggestion
const ALL_PLATFORMS = ['darwin', 'linux', 'win32'];
```

**Review by clavin — CHANGES_REQUESTED** [review]

File: modules/bot/src/github-client.ts (hunk 2 above)
Comment (erickzhao) [reply modules/bot/src/github-client.ts:170]: It's out of scope for this PR, but this draft implementation was originally added in #24 but we removed it to focus on the `/test bisect` command.

File: README.md [hunk 6]
```diff
@@ -1,20 +1,48 @@
...
+1. BugBot can bisect regressions down to a range of commits, then report back in-issue.
+1. Users and maintainers can see where the bug is reproducible.
 
-After cloning BugBot, run the following commands to set up your workspace:
-```sh
-$ yarn
-$ yarn run build
+## Usage
+
+### by Bug Reporters
```
Comment (erickzhao) [inline README.md:12]:
```suggestion
### By bug reporters
```

File: README.md [hunk 7]
```diff
@@ -1,20 +1,48 @@
...
+
+If you want to do it manually, the only requirement is that 0/1 `exitCode`. You can write it from scratch or use [electron-quickstart](https://github.com/electron/electron-quickstart)'s [`test-template` branch](https://github.com/electron/electron-quick-start/tree/test-template) as a starting point, which is the same template used by Electron Fiddle. You can clone it from the command line with `git clone --branch test-template https://github.com/electron/electron-quickstart`, then use a web browser to upload it to [gist.github.com](https://gist.github.com/).
+
+After uploading your test to gist.github.com, file an [Electron bug report](https://github.com/electron/electron/issues/new/choose). The bug report template will ask for a "Testcase Gist URL", which is where you'll provide the Gist URL.
+
+### by Maintainers
```
Comment (erickzhao) [inline README.md:22]:
```suggestion
### By maintainers
```

File: README.md [hunk 8]
```diff
@@ -1,20 +1,48 @@
 # BugBot
 
-## Development
+## tldr
 
-BugBot is split into multiple modules, connected in development via [Yarn Workspaces](https://classic.yarnpkg.com/en/docs/workspaces/) and [TypeScript Project References](https://www.typescriptlang.org/docs/handbook/project-references.html). Each module can be found within the `modules/` top-level directory.
+1. Users [file Electron bug reports](https://github.com/electron/electron/issues/new/choose) with [tests](#for-bug-reporters).
```
Comment (erickzhao) [inline README.md:5]:
```suggestion
1. Users [file Electron bug reports](https://github.com/electron/electron/issues/new/choose) with [tests](#by-bug-reporters).
```

File: README.md [hunk 9]
```diff
@@ -1,20 +1,48 @@
...
+## Usage
+
+### by Bug Reporters
+
+BugBot only needs one thing: a test that exits with `exitCode` 0 if Electron works, or nonzero if Electron has a bug.
+
+One easy way to make a test is with [Electron Fiddle](https://github.com/electron/fiddle). Its "File > New Test" has a [test template](https://github.com/electron/electron-quick-start/tree/test-template) with some [lightweight test helpers](https://github.com/electron/electron-quick-start/blob/test-template/preload.js) and its "Publish" button can upload the test to the [gist.github.com](https://gist.github.com/) [pastebin](https://en.wikipedia.org/wiki/Pastebin).
+
+If you want to do it manually, the only requirement is that 0/1 `exitCode`. You can write it from scratch or use [electron-quickstart](https://github.com/electron/electron-quickstart)'s [`test-template` branch](https://github.com/electron/electron-quick-start/tree/test-template) as a starting point, which is the same template used by Electron Fiddle. You can clone it from the command line with `git clone --branch test-template https://github.com/electron/electron-quickstart`, then use a web browser to upload it to [gist.github.com](https://gist.github.com/).
```
Comment (erickzhao) [inline README.md:18]:
```suggestion
If you want to do it manually, the only requirement is that 0/1 `exitCode`. You can write it from scratch or use [electron-quick-start](https://github.com/electron/electron-quick-start)'s [`test-template` branch](https://github.com/electron/electron-quick-start/tree/test-template) as a starting point, which is the same template used by Electron Fiddle. You can clone it from the command line with `git clone --branch test-template https://github.com/electron/electron-quick-start`, then use a web browser to upload it to [gist.github.com](https://gist.github.com/).
```

File: README.md [hunk 10]
```diff
@@ -1,20 +1,48 @@
...
+
+Much like [trop](https://github.com/electron/trop/blob/master/docs/usage.md#using-trop), you can start BugBot with issue comments. To begin bisection, add a comment that looks like this:
+
+```
+/bugbot test [gist] [platforms...] [versions...]
 ```
 
-**Note**: The latter command is required due to [a caveat with TypeScript Project References](https://www.typescriptlang.org/docs/handbook/project-references.html#caveats-for-project-references) and will hopefully be remedied automatically in the future. If you don't run `yarn run build` then you may see errors in your editor for modules that have not been built (or are not up-to-date and need to be rebuilt).
+- If no `gist` is given, use the issue body's `Testcase Gist URL`.
+- If no `platforms` are given, test on Linux, Mac, and Windows.
```
Comment (erickzhao) [inline README.md:31]:
```suggestion
- If no `platforms` are given, test on Linux, macOS, and Windows.
```

File: README.md [hunk 11]
```diff
@@ -1,20 +1,48 @@
...
+- If no `versions` are given, test the first version and latest version of each prerelease branch, of each supported branch, and of the two branches before that.
```
Comment (erickzhao) [inline README.md:32]: We can link the supported branch docs here

File: README.md [hunk 12]
```diff
@@ -1,20 +1,48 @@
 # BugBot
 
-## Development
+## tldr
```
Comment (erickzhao) [inline README.md:3]:
```suggestion
## Overview
```

**Review by erickzhao — COMMENTED** [review]
Leaving a review right now just for the docs changes

File: README.md (hunk 11 above)
Comment (ckerr) [reply README.md]: Added cb3f7fac58ab52a3bada34c0df2854914e6ea26c

File: modules/bot/src/issue-parser.ts (hunk 3 above)
Comment (ckerr) [reply modules/bot/src/issue-parser.ts:52]: Actually looks like we were passing in `undefined` below in `ret.gistId = getGistId(sections.get(TESTCASE_URL));` ... I'll fix the method signature to match how it's being used.

Comment (ckerr) [reply modules/bot/src/issue-parser.ts:52]: Updated in d63527a5b1ee6d85e926d9fb20a4f4619697b57b

File: modules/bot/src/issue-parser.ts (hunk 4 above)
Comment (ckerr) [reply modules/bot/src/issue-parser.ts:106]: > I think version ??= ... ?? ...; may be more clear than version ||= ...; version ||= ...;

No, I don't think that's right. `??=` implies we want to keep a falsy string, e.g. a version of `''`, when we really don't.

You're right though that it's strange to be allowing `undefined` for these variables without making it explicit in the types though.  I've updated that in 862aad5

File: modules/bot/src/issue-parser.ts (hunk 3 above)
Comment (clavin) [reply modules/bot/src/issue-parser.ts:52]: I still think that this function signature doesn't exactly make sense, like when would someone ever call `getGistId()` specifically with no input (or specifically with `undefined`/`null`)? Sort of a nit, but I rather modify the call site in this case to make the signature make more sense.

File: modules/bot/src/issue-parser.ts (hunk 4 above)
Comment (clavin) [reply modules/bot/src/issue-parser.ts:106]: Oh I didn't even think about `''` not being retained in this version of the code, it's just not very clear that is the intention imo.

Another thought: this code *always* evaluates the `await versions.getXyz()` and waits for it, even when it doesn't need to. I think a restructure here would be good for clarity and doing a little less work when we don't need to.

**Review by clavin — CHANGES_REQUESTED** [review]

File: modules/bot/src/issue-parser.ts (hunk 3 above)
Comment (ckerr) [reply modules/bot/src/issue-parser.ts:52]: `getGistId()` is literally called with no input / undefined / null in the code I quoted above. It passes in whatever `Map<string, string>.get(TESTCASE_URL)` has and lets the function sort it out.

File: modules/bot/src/issue-parser.ts (hunk 4 above)
Comment (ckerr) [reply modules/bot/src/issue-parser.ts:106]: That's incorrect. When `version` is truthy, the right-hand side of `||=` is not evaluated.

Comment (clavin) [reply modules/bot/src/issue-parser.ts:106]: Ah you're right, short-circuiting does work with `||=`, I just did not expect it to. Fair enough!

File: modules/bot/src/issue-parser.ts (hunk 3 above)
Comment (clavin) [reply modules/bot/src/issue-parser.ts:52]: I guess I'm still hazy about the details of how "get gist id of [nothing]" makes sense, i.e. why the handling of undefined is the responsibility of this function, but this is a non-blocker so I'm more than willing to gloss over it.

**Review by clavin — APPROVED** [review]
My remaining comments aren't blockers 👍

**Pull request merged**